graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -415,40 +327,73 @@ def _rename_and_suppress_types_and_fields(
+    if (
+        visitor.object_types_to_suppress
+        or visitor.union_types_to_suppress
+        or visitor.fields_to_suppress
+    ):
+        error_message_components = [
+            "Renamings would require further suppressions to produce a valid renamed schema."
+        ]
+        if visitor.object_types_to_suppress:
+            error_message_components.append(
+                f"The following object types have no non-suppressed fields, which is invalid: "
+                f"{sorted(visitor.object_types_to_suppress)}. To fix this, suppress the "
+                f"previously-mentioned types using the type_renamings argument of rename_schema."
+            )
+        if visitor.union_types_to_suppress:
+            error_message_components.append(
+                f"The following union types have no non-suppressed types, which is invalid: "
+                f"{sorted(visitor.union_types_to_suppress)}. To fix this, suppress the "
+                "previously-mentioned unions using the type_renamings argument of rename_schema."
+            )
+        if visitor.fields_to_suppress:
+            for object_type in visitor.fields_to_suppress:
+                error_message_components.append(f"Object type {object_type} contains: ")
+                error_message_components.extend(
+                    (
+                        f"field {field} of suppressed type "
+                        f"{visitor.fields_to_suppress[object_type][field]}, "
+                        for field in visitor.fields_to_suppress[object_type]
+                    )
+                )
+            error_message_components.append(
+                "A schema containing a field that is of a nonexistent type is invalid. To fix "
+                "this, suppress the previously-mentioned fields using the field_renamings argument "
+                "of rename_schema."
+            )
+        error_message_components.append(
+            "Note that adding suppressions may lead to other types, fields, etc. requiring "
+            "suppression so you may need to iterate on this before getting a legal schema."
+        )
+        raise CascadingSuppressionError("\n".join(error_message_components))
     if visitor.invalid_type_names or visitor.invalid_field_names:
-        explanation = (
+        error_message_components = [
             "Applying the renaming would involve names that are not valid, non-reserved "
             "GraphQL names. Valid, non-reserved GraphQL names must consist of only alphanumeric "
             "characters and underscores, must not start with a numeric character, and must not "
             "start with double underscores."
-        )
-        invalid_type_names_message = None
+        ]
         if visitor.invalid_type_names:
             sorted_invalid_type_names = sorted(visitor.invalid_type_names.items())
-            invalid_type_names_message = (
+            error_message_components.append(
```

Comment:
Needing to make the changes here surprised me. For some reason, before adding the `CascadingSuppressionError` error message above, `mypy` was fine with setting these temporary strings like `invalid_type_names_message` to `None`.

However, after adding the seemingly-unrelated `CascadingSuppressionError` code, `mypy` raised an error here because we call `"\n".join()` on a list of strings and it pointed out that initializing them to `None` meant trying to join items of `Optional[str]` type rather than `str`. Is there a known reason why it didn't catch this earlier, or is this likely a bug in `mypy` to check out?

Reply:
Likely shortcoming of mypy of some sort. Likely not worth tracking down, in terms of cost-benefit.